runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,728 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+      {
+      deps->addPostCondition(fieldClassReg, TR::RealRegister::AssignAny);
+      }
+
+   // Generates the fieldwatch reporting call.
+   generateReportFieldAccessOutlinedInstructions(node, cFlowRegionEnd, dataSnippet, isWrite, deps, cg);
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, cFlowRegionEnd, deps);
+
+   if (isResolved)
+      {
+      cg->stopUsingRegister(fieldClassReg);
+      }
+   cg->stopUsingRegister(fieldClassFlags);
+   }
+
+/*
+ * Generate instruction sequence for instance field access report.
+ */
+void indirectReadWriteBarrierHelperForFieldWatch(TR::Node *node, J9Method *owningMethod, int32_t bcIndex, bool isWrite, TR::CodeGenerator *cg)
+   {
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   J9JITWatchedInstanceFieldData instanceFieldDataBlock;
+
+   instanceFieldDataBlock.method = owningMethod;
+   instanceFieldDataBlock.location = bcIndex;
+
+   if (isResolved)
+      {
+      instanceFieldDataBlock.offset = (UDATA) (node->getSymbolReference()->getOffset() - TR::Compiler->om.objectHeaderSizeInBytes());
+      }
+   else
+      {
+      // If unresolved at compile time, we Initialize the offset to -1. The call to generateFillInDataBlockSequenceForUnresolvedField will
+      // generate instructions to resolve and correctly populate this field at runtime.
+      instanceFieldDataBlock.offset = (UDATA) -1;
+      }
+
+   TR::S390WritableDataSnippet *dataSnippet = (TR::S390WritableDataSnippet*)cg->CreateConstant(node, &instanceFieldDataBlock, 1 << (int) (log2(sizeof(J9JITWatchedInstanceFieldData)-1)+1), true);
+
+   if (!isResolved)
+      {
+      // Resolve and populate dataSnippet fields.
+      generateFillInDataBlockSequenceForUnresolvedField(node, owningMethod, dataSnippet, isWrite, cg);
+      }
+
+   TR::Register *objReg = cg->evaluate(node->getFirstChild());
+   TR::Register *j9classReg = cg->allocateRegister();
+   TR::Register *j9classFlags = cg->allocateRegister();
+ 
+   // Load the class of the instance object into j9classReg.
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, j9classReg, generateS390MemoryReference(objReg, (int32_t) TR::Compiler->om.offsetOfObjectVftField(), cg), NULL);
+
+   TR::LabelSymbol *cFlowRegionStart = generateLabelSymbol(cg);
+   TR::LabelSymbol *cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol *fieldReportLabel = generateLabelSymbol(cg);
+   cFlowRegionStart->setStartInternalControlFlow();
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   TR_ASSERT_FATAL(J9ClassHasWatchedFields >= SHRT_MIN  && J9ClassHasWatchedFields <= SHRT_MAX, "expect value of J9ClassHasWatchedFields to be with in 16 bits");
+
+   // Test the bit in fieldClassReg with the J9ClassHasWatchedFields constant to make sure fieldwatch is enabled.
+   auto instr = generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, j9classFlags, generateS390MemoryReference(j9classReg, ((TR_J9VMBase *)(cg->fe()))->getOffsetOfClassFlags(), cg));
```

Comment:
This should be correct. `fe()` returns a `TR_FrontEnd *`. But we want to call `getOffsetOfClassFlags` which is a member function of `TR_J9VMBase`. I believe it's legal to perform the cast - it's quite prevalent in the code base as well.

Reply:
I see this inside ClassLookahead.hpp:
```
TR_J9VMBase * fej9()              {return (TR_J9VMBase *)_fe; }
```
Were you referring to using this?

Reply:
@dchopra001 sorry I missed the parenthesis at the very start. I was reading it as if the cast was being applied to the result of `getOffsetOfClassFlags()` which is why I thought it was odd. However yes we can simply use `fej9()` here to avoid the cast. This would be a good change.

Reply:
The `fej9()` query is redundant across the codebase. This is another widespread problem. I've opened an issue to track this here:
https://github.com/eclipse/openj9/issues/5259